### ISSUE TITLE
Enable XCM assets on Pendulum

### ISSUE THREAD

**Issue by annatekl** [pendulum-chain/pendulum#296]
### Context
In order to move an asset from a parachain to Pendulum, the asset needs to be enabled, and opening an HRMP channel wouldn't be enough. 
e.g. [Enabling USDT on Pendulum from Asset Hub](https://github.com/pendulum-chain/pendulum/pull/181)

### Requirement
Based on the recent HRMP channels that were opened, and some of the upcoming ones, purpose of this ticket is to enable the following assets on Pendulum:

| Asset | Origination | Asset ID | Decimals | Source |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| USDC  | Asset Hub | 1337 | 6 | [subscan](https://assethub-polkadot.subscan.io/assets/1337) |   
| EQD  | Equilibrium  | 6,648,164  | 9  | Telegram  |
| PDEX  | Polkadex  |  | 12  | [subscan](https://polkadex.subscan.io/system_token_detail?unique_id=PDEX)  |
| EURC  | Pendulum  |  |   |  |
| BRZ  | Moonbeam  |  | 18  | [subscan](https://moonbeam.subscan.io/token/0xd65a1872f2e2e26092a443cb86bb5d8572027e6e) |

- USDC
- EQD
- PDEX
- EURC
- BRZ

**Comment by TorstenStueber** [comment]
@annatekl Can you specify what EQ and EQD are? It looks to me like Equilibrium has only the native asset and nothing more.

**Comment by annatekl** [comment]
I did checked there whitepaper and what I can see is:

EQ tokens are stored in user account balances within the Equilibrium blockchain's storage system.
Users can query their EQ balances using storage queries compliant with Polkadot.JS storage interfaces 

https://www.npmjs.com/package/@equilab/api
Query storage
Storage queries are compliant with [Polkadot.JS storage interfaces](https://polkadot.js.org/docs/api/cookbook/storage).

Get balances from storage method is using currencyFromU64 to decode asset u64 id into token name (eg 'eq')

import { currencyFromU64, u64FromCurrency } from "@equilab/api/equilibrium";

function getBalances(api: Api) {
  return async function (account: string): Promise<{
    ok: boolean;
    lock?: string;
    balances?: Map<string, string>;
  }> {
    const accountInfo = await api._api.query.system.account(account);
    if (!accountInfo.data.isV0) return { ok: false };

    const lock = accountInfo.data.asV0.lock.toString(10);

    const balances = accountInfo.data.asV0.balance
      .toArray()
      .reduce(
        (acc, [id, balance]) =>
          acc.set(
            currencyFromU64(id),
            balance.isPositive
              ? balance.asPositive.toString(10)
              : `-${balance.asNegative.toString(10)}`,
          ),
        new Map<string, string>(),
      );

    return { ok: true, lock, balances };
  };
}
Successfull output looks like this:

{ ok: true, lock: '22000000000', balances: { 'eq' => '72000000000', 'dot' => '597056440465' } }
Balance has 1e9 decimals places and thus '72000000000' equals 72 eq.

the documentation doesn't explicitly detail the runtime storage of EQD, but seems like that EQD would be managed using mechanisms similar to those mentioned for EQ tokens

**Comment by annatekl** [comment]
confirmed by Equilibrium: 
We have Assets pallet which is a storage of asset parameters. 
- All of our assets are 9 decimals
- Balances are stored in a system pallet. 

Asset Ids may be fetched using polkadot.js and the following script (Developer -> Javascript)

function hex2a(hexx) {
    var hex = hexx.toString();
    var str = '';
    for (var i = 0; i < hex.length; i += 2)
        str += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
    return str;
}

const assets = (await api.query.eqAssets.assets()).toJSON();  

for (const asset of assets) {
  const assetSymbol = hex2a(asset.id.toString(16)).padStart(5, ' ');
  const assetId = asset.id.toString(10).padStart(13, ' ');
  console.log(`Asset: ${assetSymbol} assetId:${assetId}`);
}

**Comment by TorstenStueber** [comment]
Okay, thank you, understood.

Two things to remark:
* It seems like their parachain is closed source – at least [their open source repo](https://github.com/equilibrium-eosdt/equilibrium-substrate-chain) is very out of date (almost three years old) and does not contain this asset pallet yet. Can you ask where to find the open source repository of their chain, please @annatekl?
* This assets pallet is not the standard asset pallet but a pallet that they implemented.

Using their definition in [the previous comment](https://github.com/pendulum-chain/pendulum/issues/296#issuecomment-1671169294) I can conclude that the assets ids are as follows:
* EQ: asset id is `25,969`
* EQD: asset id is `6,648,164`

**Comment by vadaynujra** [comment]
@annatekl as mentioned by @TorstenStueber , let's consider adding tokens from parachains that we want to integrate with after Equilibrium too.

**Comment by TorstenStueber** [comment]
@vadaynujra I guess that you specifically refer to Polkadex?

**Comment by annatekl** [comment]
@TorstenStueber this is correct:

EQ: asset id is 25,969
EQD: asset id is 6,648,164

Repo [link](https://github.com/eq-lab/equilibrium)

**Comment by TorstenStueber** [comment]
Oh thanks for the link. This is really strange: the official Equilibrium website only links to [this GitHub](https://github.com/equilibrium-eosdt), which is a completely different GitHub organization and only contains their outdated parachain code.

Could be helpful to give them this hint so that they can change it.

**Comment by annatekl** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Pendulum-6408a0365cda59c57d39bee5/issues/pendulum-chain/pendulum/296?planning-poker) with Zenhub @adelarja @ashneverdawn @b-yap @ebma @TorstenStueber

**Comment by vadaynujra** [comment]
Consider adding:

1. USDC
2. EQD (already covered)
3. PDEX
4. EURC
5. New assets from spacewalk?

**Comment by annatekl** [comment]
@ebma I have added the details for the assets (EURC is going to be issued on Pendulum), PDEX is a native token on polkadex so not sure about the id and it was not on the subscan, the same for BRZ - no details

**Comment by b-yap** [comment]
equilibrium asset info found in [asset.rs](https://github.com/eq-lab/equilibrium/blob/7f58091458bbe57d695e4be7680c924f0ff0727c/eq-primitives/src/asset.rs#L126-L188)

pallet-asset instance id found in [lib.rs](https://github.com/eq-lab/equilibrium/blob/f15982af7e3113c5c7a639fe43c55bafb39b3c20/runtime/equilibrium/src/lib.rs#L2407)

para id:
<img width="525" alt="Screen Shot 2023-11-08 at 8 35 21 PM" src="https://github.com/pendulum-chain/pendulum/assets/2826165/0640c1c5-1ce9-4215-8c73-f22e5ce4bfcc">

**Comment by ebma** [comment]
For the BRZ token, the multilocation should be the following:
```
{
  parents: 1,
  interior: {
    X3: [
      { 
        Parachain: 2004
      },
      {
        PalletInstance: 110
      },
      {
        AccountKey20: {
          key: '0xD65A1872f2E2E26092A443CB86bb5d8572027E6E'
        }
      }
    ]
  }
}

```
This is documented [here](https://docs.moonbeam.network/builders/interoperability/xcm/xc-registration/assets/#register-local-xc20), by inserting the address of [this](https://moonbeam.subscan.io/token/0xd65a1872f2e2e26092a443cb86bb5d8572027e6e) smart contract into the respective field.

**Comment by b-yap** [comment]
need Polkadex.

**Comment by annatekl** [comment]
@b-yap do you need asset id of polkadex?

**Comment by b-yap** [comment]
@annatekl yes the assetId of PDEX, or similar to what Marcel shared; which is as `AccountKey20`. 

Though I think the [pallet instance is 25](https://github.com/Polkadex-Substrate/Polkadex/blob/192cfdf002e1ba58bbbf63c837aa94c8d68508cd/runtimes/parachain/src/lib.rs#L581) but please correct me if I'm wrong.

**Comment by ebma** [comment]
@b-yap I found that they use a crate called xcm-helper which defines the conversion functions e.g. [here](https://github.com/Polkadex-Substrate/Polkadex/blob/192cfdf002e1ba58bbbf63c837aa94c8d68508cd/pallets/xcm-helper/src/lib.rs#L636). But I'm still not sure what the MultiLocation of their native asset would look like. 
@pendulum-chain/product if we have some communication channel with them, can you please ask Polkadex what the `MultiLocation` of their native asset looks like?

**Comment by annatekl** [comment]
@b-yap @ebma I asked via telegram, they provided me with Multilocation: {parents: 1, interior: {x1:{parachain: 2040}}}

**Comment by b-yap** [comment]
for `EURC`, what are the ff. tickets that need to be resolved first? 
https://github.com/pendulum-chain/tasks/issues/87
https://github.com/pendulum-chain/tasks/issues/128
https://github.com/pendulum-chain/pendulum/issues/310
https://github.com/pendulum-chain/pendulum/issues/331
https://github.com/pendulum-chain/pendulum/issues/334

^ I'm naming a few, but I'm not quite sure
@annatekl

**Comment by ebma** [comment]
The only real blocker I see is https://github.com/pendulum-chain/pendulum/pull/342. Once that is merged, you should be able to configure EURC. Though we did not yet specify how we want the MultiLocation of our new `CurrencyId::Token` types to look like externally. We would also need to discuss this in the tech team. Do you have a suggestion for the MultiLocation @b-yap?

**Comment by b-yap** [comment]
@ebma I checked all the variants in `Junction`, and what makes sense (to me) is the `GeneralIndex`. 
We could follow how Equilibrium does it, with `Asset(<value>)` where `<value>` came from [fn from_bytes()](https://github.com/eq-lab/equilibrium/blob/7f58091458bbe57d695e4be7680c924f0ff0727c/eq-primitives/src/asset.rs#L235)

So it will be
```rust
MultiLocation {
	parents: 1,
	interior: X3(
		Parachain(id),
		PalletInstance(10),
		GeneralIndex(from_bytes(b"EURC"))
	),
}
```
and
```rust
MultiLocation {
	parents: 0,
	interior: X2(
		PalletInstance(10),
		GeneralIndex(from_bytes(b"EURC"))
	),
}
```

**Comment by ebma** [comment]
Hmm interesting. I would put EURC as `Token(0)` but we can use anything. 
For the MultiLocation, I'm not sure. First of all, we should probably change the number in `PalletInstance` to point to the `Tokens` pallet as `10` points to our `Balances` pallet which only makes sense for our native token.

Now that we point to our `Tokens` pallet, we still need to map the asset somehow to our `CurrencyId` enum. We should probably already think about how we can map _any_ type of our `CurrencyId` enum to a `MultiLocation` so we should also consider Stellar-type assets in addition to the `CurrencyId::Token()`. 

We could use a combination of [GeneralIndex](https://github.com/paritytech/polkadot-sdk/blob/f4bb17cc863166622d407cadc16c467d4bf1439e/polkadot/xcm/src/v2/junction.rs#L64) and [GeneralKey](https://github.com/paritytech/polkadot-sdk/blob/f4bb17cc863166622d407cadc16c467d4bf1439e/polkadot/xcm/src/v2/junction.rs#L70) with the index pointing to the index of the type in our `CurrencyId` enum, with Stellar being index `2` and `Token` being index 4, and then putting the actual specifier into the GeneralKey. 

WDYT @b-yap and also @pendulum-chain/devs?

**Comment by gianfra-t** [comment]
I think the solution of using the junction `GeneralIndex` and `GeneralKey` would very well. Because even if we later would like to use `GeneralIndex` for other MultiLocations that do not fall into the category of the `CurrencyId` enum, we still have a very large space (u128) to represent them.
I imagine that for Stellar assets the key would be something like `GeneralKey([from_bytes(b"USDT", b"Issuer..."].concat())` right?

**Comment by b-yap** [comment]
@gianfra-t  The `GeneralKey` has a max data length of 32. `b"USDT" + b"Issuer..."` might be difficult. 
Is it possible to have multiple `GeneralKey` junctions to exist? Although usage of the `GeneralKey` as pointed in the comment, must be avoided (if possible).

@ebma I'm using the mod `latest`, so this is I think the link of [GeneralKey](https://github.com/paritytech/polkadot-sdk/blob/f4bb17cc863166622d407cadc16c467d4bf1439e/polkadot/xcm/src/v3/junction.rs#L311).
Is there a preference to use `v2` rather than `v3` immediately?

```rust
MultiLocation {
	parents: 0,
	interior: X3(
		PalletInstance(53), // 53 is Token pallet
		GeneralIndex(4),
		GeneralKey{ ?? }
	),
}
```
Let's say`EURC` is `Token(0)`; will the `GeneralKey` represent the "0" ? It makes sense for it to be 0.

**Comment by ebma** [comment]
> Is it possible to have multiple GeneralKey junctions to exist?

Yes we can use those multiple times.

> Although usage of the GeneralKey as pointed in the comment, must be avoided (if possible).

I don't think we need to avoid it, the comment is probably meant to make people try using more specific items if possible. But there is no problem if we don't.

> Is there a preference to use v2 rather than v3 immediately?

Not really, I just chose the first that I found 😅 

> Let's say EURC is Token(0); will the GeneralKey represent the "0" ? It makes sense for it to be 0.

Yeah we could do that.

> The GeneralKey has a max data length of 32. b"USDT" + b"Issuer..." might be difficult.

So for Stellar assets we could also use two consecutive `GeneralKey`s like 
```
interior: X3(
		PalletInstance(53), // 53 is Token pallet
		GeneralIndex(2),
		GeneralKey{ length: 4, // or 12 if it is an AlphaNum12
                                       data: code  
                 },
                 GeneralKey{ length: 32,
                                       data: issuer  
                 },

	),
```

**Comment by TorstenStueber** [comment]
Here are a few unrelated thoughts.

1. @b-yap pointed out that using the value `Token(0)` to stand for EURC is not elegant. I agree and this was actually my misguiding input. However, I assume that this now already rolled out on Amplitude (?) and therefore too hard to change again (?)

2. Whether we use V2 or V3 should not matter. Usually V3 is just an extension of V2 and everything we an define as V2 would automatically internally converted to V3 if applicable. This conversion is transparent and we don't need to care about this (take this advice with care).

3. I agree with Marcel's previous comment. I think it's better to be explicit and therefore prefer to use multiple GeneralIndex and GeneralKey combinations.

**Comment by b-yap** [comment]
@TorstenStueber @ebma 
> However, I assume that this now already rolled out on Amplitude (?) and therefore too hard to change again (?)

 -> https://github.com/pendulum-chain/pendulum/pull/342 merged just recently.

**Comment by TorstenStueber** [comment]
Oh, merged so recently that it is not part of the Amplitude release yet so that we could still change this.

**Comment by ebma** [comment]
I would propose the following. Considering our current [CurrencyId](https://github.com/pendulum-chain/spacewalk/blob/82f9e432490b85234f16538b996c589e6af4146e/primitives/src/lib.rs#L493) enum
```
pub enum CurrencyId {
	Native = 0_u8,
	XCM(u8),
	Stellar(Asset),
	ZenlinkLPToken(u8, u8, u8, u8),
	Token(u64),
}
```
## Mappings
### Native
We actually define a different `MultiLocation` for our `Native` asset, referring to the `Balances` pallet.
```
MultiLocation {
	parents: 1,
	interior: X2(
		Parachain(id),
		PalletInstance(10),
	),
}
```

### XCM
As these are all foreign assets, their MultiLocation is defined by the source chain.

### Stellar
For `StellarNative` we might just leave out 

```
X2(
		PalletInstance(53), // 53 is Token pallet
		GeneralIndex(2),
	),
```

For `Stellar::AlphaNum4`

```
X4(
		PalletInstance(53), // 53 is Token pallet
		GeneralIndex(2),
		GeneralKey{ length: 4,  data: code }, // eg. b"USDC"
                 GeneralKey{ length: 32, data: issuer  }, // the raw binary of the public key
	),
```

For `Stellar::AlphaNum12`
```
X4(
		PalletInstance(53), // 53 is Token pallet
		GeneralIndex(2),
		GeneralKey{ length: 12, data: code  }, // eg b"USDC\0\0\0\0\0\0\0\0
                 GeneralKey{ length: 32, data: issuer },
	),
```

### ZenlinkLPToken
We need to encode the four `u8`s 
```
X4(
		PalletInstance(53), // 53 is Token pallet
		GeneralIndex(3),
		GeneralKey{ length: 4, data: [u8, u8, u8, u8]  }, // We can just concatenate them
	),
```

### Token

```
X4(
		PalletInstance(53), // 53 is Token pallet
		GeneralIndex(4),
		GeneralIndex(index), // The index of our token in the token enum

	),
```

**Comment by TorstenStueber** [comment]
@ebma Great, I like it.
